### PR TITLE
FOGL-4347 API changes when APT operations are in progress 

### PR DESCRIPTION
**Pull request by ashish-jabble**
**Pending items**

- [x] POST install endpoint
- [x] DELETE endpoint
- [x] UPDATE endpoint
- [x] unit tests
- [x] Review TODO & FIXME comments
- [x] Sqlite upgrade & downgrade scripts for packages table
- [x] with Postgres init.sql changes, upgrade/downgrade scripts
- [x] RPM testing
- [x] Package system test needs to be updated
- [x] refresh cache for get available plugins when purge the package
- [x] Service POST action=install
- [x] Service PUT

**NOTE** - GUI functionality also needs to be changed. Related PR https://github.com/fledge-iot/fledge-gui/pull/38
                   With white-labelled we might also change the FogMan code by calling these API's

Packages available at http://archives.fledge-iot.org/fixes/FOGL-4347/
Similar for white-labelled

Jenkins run `Fledge%20Nightly%20Jobs/job/fledge_nightly_build_ub1804/421/` - **All system & unit tests are PASSED**

Examples:
```
$ curl -sX POST http://localhost:8081/fledge/plugins -d '{"format":"repository", "name": "fledge-south-benchmark"}' | jq
{
  "message": "Plugin installation started.",
  "id": "73f5179c-805b-4265-b69a-9f481959d635",
  "statusLink": "fledge/package/install/status?id=73f5179c-805b-4265-b69a-9f481959d635"
}

Track its status via below endpoint (basically POLL)
Try-1
$ curl -sX GET http://localhost:8081/fledge/package/install/status?id=73f5179c-805b-4265-b69a-9f481959d635 | jq
{
  "packageStatus": [
    {
      "id": "73f5179c-805b-4265-b69a-9f481959d635",
      "name": "fledge-south-benchmark",
      "action": "install",
      "status": "in-progress",
      "logFileURI": ""
    }
  ]
}

Try-2

$ curl -sX GET http://localhost:8081/fledge/package/install/status?id=73f5179c-805b-4265-b69a-9f481959d635 | jq
{
  "packageStatus": [
    {
      "id": "73f5179c-805b-4265-b69a-9f481959d635",
      "name": "fledge-south-benchmark",
      "action": "install",
      "status": "success",
      "logFileURI": "log/201006-15-32-57-fledge-south-modbus-install.log"
    }
  ]
}


$ curl -sX DELETE http://localhost:8081/fledge/plugins/south/Benchmark | jq
{
  "message": {
    "message": "Benchmark plugin purge started.",
    "id": "b597439d-e3ca-4cad-99a3-00b36a9b1cd8",
    "statusLink": "fledge/package/purge/status?id=b597439d-e3ca-4cad-99a3-00b36a9b1cd8"
  }
}
aj@aj-ThinkPad-E450 ~ $ curl -sX GET http://localhost:8081/fledge/package/purge/status?id=b597439d-e3ca-4cad-99a3-00b36a9b1cd8 | jq
{
  "packageStatus": [
    {
      "id": "b597439d-e3ca-4cad-99a3-00b36a9b1cd8",
      "name": "fledge-south-benchmark",
      "action": "purge",
      "status": "success",
      "logFileURI": "log/201006-15-41-46-fledge-south-benchmark-remove.log"
    }
  ]
}

```

### CODE REVIEW

**Review by MarkRiddoch — COMMENTED** [review]
Looks like a good approach tome. We will need to update the GUI and FogMan to work with the new semantics of the API.

File: python/fledge/services/core/api/plugins/install.py [hunk 1]
```diff
@@ -279,8 +330,12 @@ async def install_package_from_repo(name: str, pkg_mgt: str, version: str) -> tu
...
                 ret_code = os.system(cmd + " > {} 2>&1".format(stdout_file_path))
                 if ret_code != 0:
-                    raise PackageError(link)
-                pkg_cache_mgr['upgrade']['last_accessed_time'] = now
+                    # Update record in Packages table for given uid only in case of APT upgrade fails
+                    payload = PayloadBuilder().SET(status=ret_code, log_file_uri=link).WHERE(['id', '=', uid]).payload()
+                    loop.run_until_complete(storage.update_tbl("packages", payload))
+                    return
+                else:
+                    pkg_cache_mgr['upgrade']['last_accessed_time'] = now
```
Comment (praveen-garg) [inline python/fledge/services/core/api/plugins/install.py:338]: Update this via cache update route; else it will not be effective in core process.

Comment (ashish-jabble) [reply python/fledge/services/core/api/plugins/install.py:338]: We have never called upgrade in any of case. Upgrade case only in core process as of now.

Comment (praveen-garg) [reply python/fledge/services/core/api/plugins/install.py:338]: I see `cmd = "sudo {} -y upgrade".format(pkg_mgt) if pkg_mgt == 'apt' else "sudo {} -y update".format(pkg_mgt)` in `if duration_in_sec` block

**Review by MarkRiddoch — APPROVED** [review]

File: tests/unit/python/fledge/services/core/api/plugins/test_install.py [hunk 2]
```diff
@@ -279,27 +271,102 @@ async def async_mock(return_value):
...
+                        "version": "1.8.1", "installedDirectory": "south/ModbusC", "packageName": plugin_name}]
+        msg = '{} package is already installed'.format(plugin_name)
+        storage_client_mock = MagicMock(StorageClientAsync)
+        with patch.object(connect, 'get_storage_async', return_value=storage_client_mock):
+            with patch.object(storage_client_mock, 'query_tbl_with_payload',
+                              return_value=async_mock({'count': 0, 'rows': []})) as query_tbl_patch:
+                with patch.object(PluginDiscovery, 'get_plugins_installed', return_value=plugin_list
+                                  ) as plugin_list_patch:
+                    resp = await client.post('/fledge/plugins', data=json.dumps(param))
+                    assert 304 == resp.status
```
Comment (praveen-garg) [inline tests/unit/python/fledge/services/core/api/plugins/test_install.py:367]: This should be 400/ 409 / 422 *not 304*

**Review by MarkRiddoch — APPROVED** [review]

**Comment by ashish-jabble** [comment]
@MarkRiddoch This is done from my side.

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**